src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -3,88 +3,146 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections;
+using System.Collections.Generic;
 using System.Collections.Specialized;
 
 namespace System.Net
 {
     internal class HeaderInfoTable
     {
-        private static Hashtable HeaderHashTable;
-        private static HeaderInfo UnknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false);
+        private static readonly Func<string, string[]> s_singleParser = value => new[] { value };
+        private static readonly Func<string, string[]> s_multiParser = value => ParseValueHelper(value, isSetCookie: false);
+        private static readonly Func<string, string[]> s_setCookieParser = value => ParseValueHelper(value, isSetCookie: true);
+        private static readonly HeaderInfo s_unknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, s_singleParser);
+        private static readonly Hashtable s_headerHashTable;
+
+        private static string[] ParseValueHelper(string value, bool isSetCookie)
+        {
+            List<string> tempStringCollection = new List<string>();
+
+            bool inquote = false;
+            int chIndex = 0;
+            char[] vp = new char[value.Length];
+            string singleValue;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                {
+                    inquote = !inquote;
+                }
+                else if ((value[i] == ',') && !inquote)
+                {
+                    singleValue = new string(vp, 0, chIndex);
+                    if (!isSetCookie || !IsDuringExpiresAttributeParsing(singleValue))
+                    {
+                        tempStringCollection.Add(singleValue.Trim());
+                        chIndex = 0;
+                        continue;
+                    }
+                }
+                vp[chIndex++] = value[i];
+            }
+
+            // Now add the last of the header values to the stringtable.
+            if (chIndex != 0) {
+                singleValue = new string(vp, 0, chIndex);
+                tempStringCollection.Add(singleValue.Trim());
```

Comment:
This isn't affected by isSetCookie or IsDuringExpiresAttributeParsing as in the other cases?

Reply:
No.

The purpose for isSetCookie & IsDuringExpiresAttributeParsing is that to preserve the integrity for the "Expires" attribute. (Because this attribute contains one comma, we need to make sure it's not splitted)

If the last header value is "Expires" attribute for Set-Cookie header, previous check has ensure the integrity. We can safely add it to the string table.

If it's not, it's the usual case. Add it to the string table.